### ISSUE TITLE
The menu item 'Save As/Export' doesn't do anything

### ISSUE THREAD

**Issue by yurivict** [DISTRHO/Cardinal#135]
Usually 'Save As' action is expected to ask the user about file path.

**Comment by falkTX** [comment]
Yes. This requires dbus with a desktop-portal implementation on the desktop environment.
This is not the case for you?

The open file still works, correct?

**Comment by yurivict** [comment]
A better behavior here is to determine that desktop-portal isn't running and tell this to the user is a pop-up box.

open works.

**Comment by falkTX** [comment]
I am putting this in the FAQ, but it is expected that eventually all desktop environments are going to support it.
https://github.com/flatpak/xdg-desktop-portal lists several backends already made. I have checked with some people and it works on Linux with Gnome, KDE and LXDE.

**Comment by yurivict** [comment]
But if it doesn't work for some reason - it's better to explain to the user why the action can't be completed.

**Comment by falkTX** [comment]
I can show an error, but knowing exactly why would not break the API which I am not very keen on doing.
A generic warning is pointless, so I guess it is best for us to try to figure this out then.

Do you have `qdbusviewer` ?
It should help to debug this, as it can nicely traverse/introspect all the active DBus paths and methods.

**Comment by yurivict** [comment]
It doesn't show ```org.freedesktop.portal.Desktop```. I am on XFCE4.

So it would seem like you could try to look up this name in DBus and if not found tell something like this to the user:
> Desktop-portal doesn't seem to be running on your system. The requested action can't be completed.

**Comment by falkTX** [comment]
> It doesn't show `org.freedesktop.portal.Desktop`. I am on XFCE4.

Are you able to install some package to get the gtk implementation?

**Comment by dromer** [comment]
On our Devuan machine at the hackerspace I have to run `/usr/libexec/xdg-desktop-portal` manually, which is quite annoying.

However programs like Evince are able to either trigger the start of `xdg-desktop-portal` (on my debian machines) or access the file dialog themselves without the portal running (devuan), so it might be good to check the Evince source to see how they are doing this.

For the user it is certainly quite annoying to have to debug and dig for work-arounds like this, just to get basic save functions to work.

**Comment by bambax** [comment]
My Windows username has non-ascii characters. The save as dialog points to the wrong user directory. (Win7, Reaper, LV2).

**Comment by falkTX** [comment]
> My Windows username has non-ascii characters. The save as dialog points to the wrong user directory. (Win7, Reaper, LV2).

that has nothing to do with this, please open a separate ticket

**Comment by bambax** [comment]
> > My Windows username has non-ascii characters. The save as dialog points to the wrong user directory. (Win7, Reaper, LV2).
> 
> that has nothing to do with this, please open a separate ticket

Sorry. Opened #137

**Comment by spectromas** [comment]
Same thing for me, save as does nothing in standalone or in the vst2. I have qdbusviewer but I'm not sure what I should be looking for? I cannot see `org.freedesktop.portal.Desktop` listed.

**Comment by spectromas** [comment]
I have that installed and works elsewhere (I can change the implementation in firefox and Rack for example).
![Screenshot_20220215_095113](https://user-images.githubusercontent.com/6443372/154037282-08286556-0b34-4b07-b092-aee9909df5d5.png)

**Comment by falkTX** [comment]
Rack doesnt use desktop portal.
And even while installed, doesn't mean it is active. As you stated:

> I cannot see `org.freedesktop.portal.Desktop` listed

I dont know why this is the case, but it is the cause of the issue.

**Comment by spectromas** [comment]
For me Rack uses the KDE file dialog and changes to the GTK one if I don't have `GTK_USE_PORTAL=1` in /etc/environment, just like firefox. 

Strangely the save as is working now. I disabled `GTK_USE_PORTAL=1`, rebooted and it made no difference so I reverted, logged out and logged back in and now it is working and I can see the `org.freedesktop.portal.Desktop` entry.

**Comment by yurivict** [comment]
FYI: The [Nuklear](https://github.com/Immediate-Mode-UI/Nuklear) single-header library allows to show the file dialog using only OpenGL. There's no need to use DBus.

**Comment by falkTX** [comment]
DPF can do it too with sofd, but only for opening not for saving.
text input is not an easy thing to do.

**Comment by falkTX** [comment]
Yes I know. but works quite differently, I cant simply copy&paste that code, it is not how it works.

**Comment by yurivict** [comment]
Can't File/Save dialog be reduced to just one function?

**Comment by falkTX** [comment]
No, we dont have such native dialogs on linux, we always need custom code.
This is clearly not simple https://github.com/DISTRHO/DPF/blob/develop/distrho/extra/sofd/libsofd.c

**Comment by dromer** [comment]
Works for me on debian bullseye!

Should take care of the majority of cases at least (I expect this to still fail on devuan which has a broken xdg-desktop-portal package).

**Comment by maxigaz** [comment]
The "Save as/Export..." dialogue doesn't appear for me in Cardinal 22.04 (standalone or as a plugin inside Carla). The "Open/Import..." dialogue works, though.

I'm running Arch Linux with i3, so without a full-fledged desktop environment. I've tried installing `xdg-desktop-portal-gtk` and `xdg-desktop-portal-gnome`, but I get the same results. Do I need to install anything else?

**Comment by falkTX** [comment]
I think those are the ones to install, but you need to be running `xdg-desktop-portal` in order for the dbus service to be available.

**Comment by maxigaz** [comment]
I've found that there are systemd service files for both `xdg-desktop-portal-gtk` and `xdg-desktop-portal-gnome`. I can run them with `systemctl` (e.g. `systemctl xdg-desktop-portal-gtk --user`), but no change.

Running `systemctl enable` shows a message that these unit files are not meant to be enabled manually.

I've even tried running `/usr/lib/xdg-desktop-portal-gtk`, what the unit file points to, from a terminal. Still no dialogue for saving.

**Comment by falkTX** [comment]
I think the right command for the start is
```
systemctl --user --start xdg-desktop-portal-gtk
```

but well, quite a few things are moving into the desktop portal direction, special sandboxed applications like snap and flathub.
this ends up being a system problem.

if you have suggestions on how to make this better, I am all ears, but the approach to the file dialogs is not really changing in meaningful ways.

**Comment by maxigaz** [comment]
After some more testing, here's what else I've found:

* Running Firefox automatically triggers the portal service, so after that no manual action is required.
* If the service is not running (i.e. Firefox hasn't been run at least once during a login session), I need to select File > Save/Export... inside Cardinal *twice*. (The service is started the first time, and the dialogue appears the second time.)
* `systemctl enable xdg-desktop-portal --user` can be used to run the service manually.
* At least under Arch Linux and i3, having `xdg-desktop-portal-gnome` installed in addition to `xdg-desktop-portal` and `xdg-desktop-portal-gtk` seems to introduce some issues: The dialogue doesn't follow the system theme and typing every other character in the filename field shifts focus to the search field, which makes it borderline unusable. Uninstalling `xdg-desktop-portal-gnome` solved it all for me.

**Comment by maxigaz** [comment]
> You sure you are running 22.04?

Yes, I am.

![cardinal-2022-04-11_16-46-30](https://user-images.githubusercontent.com/10419911/162765776-d944a25e-4abb-40db-982f-610db8128867.png)

**Comment by coax75ohm** [comment]
I'm running 22.12 installed via KXStudio onto Raspberry Pi OS and encountered this issue.

`ps` shows that `xdg-desktop-portal` is running.

As a workaround, I've found that Carla can be used to save state with the plugin version.

**Comment by vacuumbeef** [comment]
@maxigaz as you said above, "Running systemctl enable shows a message that these unit files are not meant to be enabled manually."
And then you say "systemctl enable xdg-desktop-portal --user can be used to run the service manually."
I don't get it.
I have xdg-desktop-portal and xdg-desktop-portal-gtk installed, and it tells me "This means they are not meant to be enabled using systemctl".
I'm on Arch with dwm, Cardinal 22.12

**Comment by maxigaz** [comment]
@vacuumbeef I can't reproduce the original issue anymore.

I've just tested Cardinal 22.12 running inside Carla in an Arch VM. The Save As/Export dialogue appears correctly without having to start or enable any systemd services (or even Firefox) by hand.

> I have xdg-desktop-portal and xdg-desktop-portal-gtk installed, and it tells me "This means they are not meant to be enabled using systemctl".

I can't reproduce that one either.

**Comment by coax75ohm** [comment]
Finally figured out the issue I was having is that the base xdg-desktop-portal package didn't enable the FileChooser portal.  After installing xdg-desktop-portal-gtk it works.

**Comment by marrs** [comment]
Same issue here using Cardinal Standalone v23.07, installed via from the extra repository on Arch. I'm also not running a mainstream desktop environment.

As advised above, I've installed `xdg-desktop-portal` and `xdg-desktop-portal-gtk` but this hasn't helped.  `gtk` seemed reasonable to me given I have that library installed.  As mentioned by another user, systemctl will not let me enable this manually.

I also tried enabling the env var for `XDG_CURRENT_DESKTOP` in case the lack of this env var was preventing `xdg-desktop-portal` from selecting the appropriate backend, but so far this hasn't helped.  I used the following commands to achieve this:
```
export XDG_CURRENT_DESKTOP=X-Generic
systemctl --user import-environment  XDG_CURRENT_DESKTOP
```

I guess the next step will be to inspect the DBus logs.  I did have a look in `qdbusviewer` for `org.freedesktop.portal.Desktop` but it's not present.

> I am putting this in the FAQ, but it is expected that eventually all desktop environments are going to support it.

Maybe so, but many users don't use "desktop environments" of the kind you're referring to.  My desktop environment was built by me, and this is quite a common scenario amongst Linux users.  I'd expect any Linux app to be able to be desktop agnostic.

> I can show an error, but knowing exactly why would break the API which I am not very keen on doing.
A generic warning is pointless, so I guess it is best for us to try to figure this out then.

The expectation is that if there is no warning then it worked.  Initially, I thought Cardinal was being clever and saving my patch to a default location, because it didn't show me a save dialogue and it didn't complain.  It was only after I restarted the app that I realised I was wrong.

Even a "failed to save" message is infinitely preferable to making the user guess what happened.

**Comment by falkTX** [comment]
I understand, but from my experience this is not a typical setup.

And the item is already on the FAQ https://github.com/DISTRHO/Cardinal/blob/main/docs/FAQ.md#on-bsdlinuxx11-the-menu-item-save-asexport-does-nothing
With a clear command for enabling the desktop portal.

```
systemctl enable xdg-desktop-portal --user --now
```

This is the same setup as used in flatpaks and web browsers (plus any electron-based tools too eventually), so with time it is just going to get more commonplace

**Comment by marrs** [comment]
> I understand, but from my experience this is not a typical setup.

I use Electron based tools and a variety of web browsers, including Chrome, Brave, and Firefox, and they all access the filesystem just fine.  I can't claim to understand the nature of the problem, but this is the first app I've ever experienced it with.

In any case, I'll have another go at debugging tonight.

**Comment by falkTX** [comment]
that is very curious then, I expected them to not work if the desktop portal is not available, specially firefox.

cardinal uses the same exact dbus approach as these other tools, maybe some of them have fallbacks for gtk dialogs (which is not suitable for cardinal)

maybe a wayland vs x11 thing? in case you are wayland, I did not test Cardinal in there much.

**Comment by marrs** [comment]
I worked it out.   I used dbus-monitor to observe the call to dbus and, between that and journalctl, I discovered that my monitor was not being recognised.

It turns out I needed to run `dbus-update-activation-environment --systemd DBUS_SESSION_BUS_ADDRESS DISPLAY XAUTHORITY`, which I added to my `~/.xinitrc`, which will hopefully set it every time I start X.

I still think the user experience should be better on X, but at least this info exists for anyone else who has the same issue.

**Comment by Yngwarr** [comment]
> I understand, but from my experience this is not a typical setup.

It's more typical than you'd expect. :)

I run Cardinal 24.05 Arch with i3wm and x11 and neither open or save dialogs work. The command from FAQ clearly states that it shall not be called this way and it doesn't help. Thank you, @marrs, your solution helped!
